### PR TITLE
SAI PTF 1: Proposal

### PR DESCRIPTION
**Pull request by pkochmanska**

### CODE REVIEW

File: meta/gensairpc.pl [hunk 1]
```diff
@@ -0,0 +1,605 @@
+#!/usr/bin/env perl
+#
+# Copyright (c) 2014 Microsoft Open Technologies, Inc.
```
Comment (aljer) [inline meta/gensairpc.pl:3]: 2021

File: doc/SAI-Proposal-SAI-PTF.md [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+
+![autogeneration gensairpc](figures/autogeneration_gensairpc.png "Figure 3: Autogeneration - gensairpc")
+__Figure 3: Autogeneration - parsing__
+
+Autogeneration when upgrading SAI:
+
+```
+cd <root>/meta
+make
+./gensairpc.pl --clean --no-meta-build
```
Comment (richardyu-ms) [inline doc/SAI-Proposal-SAI-PTF.md:70]: seems this parameter doesn't existing in perl

File: meta/templates/sai_adapter.py.tt [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+        [%- PROCESS catch_exception %]
+[%- END -%]
+
+[%- ######################################################################## -%]
+
+[%- ######################################################################## -%]
+
+[%- BLOCK catch_exception -%]
+    except sai_thrift_exception as e:
+        status = e.status
```
Comment (richardyu-ms) [inline meta/templates/sai_adapter.py.tt:165]: how about generating those exception handlers into a utils method that can be used by the user?

File: meta/gensairpc.README [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    In order to generate debug files and avoid cleanup, use --dbg or --dump
+    flags. Use --experimental to generate extensions and experimental code.
+
+    For more information, use --help:
+
+      ./gensairpc.pl --help
+
+DEPENDENCIES
+    Ubuntu packages:
```
Comment (richardyu-ms) [inline meta/gensairpc.README:27]: is there any place that mentioned 'graphviz' as a dependency for 'dot' which required in Makefile 'BINS = doxygen perl dot'
Not sure, if libmoose-perl needed

File: meta/templates/sai_adapter.py.tt (hunk 3 above)
Comment (aljer) [reply meta/templates/sai_adapter.py.tt:165]: Not sure I understand your point of view. This block simply surround each RPC server function call in sai_adapter with try-except clause. How would you like to turn it into a method? Secondly, do you mean to put this into sai_utils file that we want to add here: https://github.com/opencomputeproject/SAI/pull/1335/files#diff-7a25e711c6b68d0df0743a168d6efcdb1258b18e7062942eefd507eec767d9cb or some other? sai_utils file is not intended to be autogenerated. It is written manually.

File: doc/SAI-Proposal-SAI-PTF.md [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+- default_1q_bridge
+- acl_stage_ingress
+- acl_stage_egress
+- number_of_active_ports
+- port numbers stored into variables portX, where X is the number of port
+
+SaiHelper inherits from SaiHelperBase but provides additional configuration.
+The base configuration created by SaiHelper looks like this:
+
+| Name       | Vlan ID | Ports | Tagging |
```
Comment (aljer) [inline doc/SAI-Proposal-SAI-PTF.md:104]: Please update the table according to the new one we have in sai_base_test file that is to be upstreamed (https://github.com/aljer/SAI/blob/sai_ptf_infra/ptf/sai_base_test.py#L490)

**Review by kcudnik — APPROVED** [review]

File: meta/templates/sai_adapter.py.tt (hunk 3 above)
Comment (richardyu-ms) [reply meta/templates/sai_adapter.py.tt:165]: I see, there is a condition 'if CATCH_EXCEPTIONS:'.

Comment (richardyu-ms) [reply meta/templates/sai_adapter.py.tt:165]: but still, have some concerns about this try-except, first, it is not thread-safe. secondly, an exception should be happened as unexpected. why do we need to wrap that, and also request a manual check for the status, but not throw it directly and let users to handle them as their needs?

File: meta/templates/sai.thrift.tt [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+    [%- END -%]
+[% END -%]
+
+[%- ######################################################################## -%]
+
+[%- # The body of the file: -%]
+// AUTOGENERATED FILE! DO NOT EDIT
+
+namespace py sai
```
Comment (richardyu-ms) [inline meta/templates/sai.thrift.tt:233]: current namespace is switch_sai. 
saiservee.cpp depends on those generated files, like switch_sai_rpc, do we plan to use a new saiserver.cpp?
saiserver is referenced by syncd-rpc, then we need to upgrade the syncd-rpc to use the new saiserver if we want to use the new structure?

Comment (ravi861) [reply meta/templates/sai.thrift.tt:233]: Yes. The 2nd PR for SAI PTF has a new sai_rpc_frontend.cpp which uses the new namespace. I am going to create a 3rd PR which will include the Makefile.

File: meta/gensairpc.README (hunk 4 above)
Comment (aljer) [reply meta/gensairpc.README:27]: 'BINS = doxygen perl dot' is an old part of Makefile. We don't introduce 'dot' with the generator. That's why its dependencies are not listed here.

Comment (aljer) [reply meta/gensairpc.README:27]: \+ IMO we could add libmoose-perl to the list, but it is a dependency of libmoosex-aliases-perl so they should be installed together.

File: meta/templates/sai_adapter.py.tt (hunk 3 above)
Comment (pkochmanska) [reply meta/templates/sai_adapter.py.tt:165]: This is the design we established and according to that we have written all the tests, this change would mean we have to modify all the test cases we have to handle exceptions.

Comment (richardyu-ms) [reply meta/templates/sai_adapter.py.tt:165]: ok, i see

**Review by richardyu-ms — APPROVED** [review]

**Comment by ravi861** [comment]
@lguohan @rlhui Should we give more time for reviews or can we go ahead with merge?

**Comment by KrisNey-MSFT** [comment]
Hi team, per our SONiC-DASH-Test-Workgroup community meeting this morning, may we ask to accommodate SONiC-DASH (NIC) as we move forward?  @chrispsommers @rlhui @ravi861 
We are hoping to leverage SAI-Thrift for test purposes.
Thank you, @Krisney-MSFT

**Comment by ravi861** [comment]
@KrisNey-MSFT Can you please reach out to me over email ravi.vantipalli@intel.com for what you are looking for?

**Comment by KrisNey-MSFT** [comment]
Yes, will do - thank you.
Kristina



Kristina Moore, MBA, M.S., CISSP

Principal Program Manager

Azure Networking – Hardware/Switching/Performance

400G New Tech

Office:   40/4J00

Mobile:   (425) 876- 2040<tel:(425)%20876-%202040>

***@***.******@***.***>



***@***.***





________________________________

**Comment by KrisNey-MSFT** [comment]
Thanks Ravi, would you want to chat tomorrow at 1pm or so?  My #'s are below, or you could give me yours..?

***@***.***

Kristina Moore MBA, M.S., CISSP
Azure Core Principal PM
DASH

Office: 425-722-7720
Mobile: 425-876-2040
***@***.***

***@***.***

**Comment by reshmaintel** [comment]
@KrisNey-MSFT Hi Kristina, would you please include me in the SAI PTF call with Ravi/Intel. Alternately I can arrange a meeting for all of us to discuss. Thanks,

**Comment by richardyu-ms** [comment]
> @KrisNey-MSFT Hi Kristina, would you please include me in the SAI PTF call with Ravi/Intel. Alternately I can arrange a meeting for all of us to discuss. Thanks,

Me too, thanks.

**Review by rlhui — APPROVED** [review]

**Pull request merged**